### PR TITLE
iproto: introduce `IPROTO_INSERT_ARROW` request

### PR DESCRIPTION
**Pull request by Gumix**
The new request inserts into a given space the data that is encoded into Arrow IPC columnar format [^1].
At the moment the request is not supported by memtx and vinyl spaces.

[^1]: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc

Closes https://github.com/tarantool/tarantool/issues/10508
Needed for https://github.com/tarantool/tarantool-ee/issues/820

### CODE REVIEW

File: changelogs/unreleased/gh-10508-iproto_insert_arrow.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## feature/core
+
+* Introduced the new `IPROTO_INSERT_ARROW` request for batch insertions of the
+  data in Arrow format (gh-10508).
```
Comment (p7nov) [inline changelogs/unreleased/gh-10508-iproto_insert_arrow.md:4]:
```suggestion
* Introduced the new `IPROTO_INSERT_ARROW` request for batch data insertions
  in the Apache Arrow format (gh-10508).
```

**Review by p7nov — APPROVED** [review]

File: src/box/arrow_ipc.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		 struct arrow_ipc_buffer *ipc_buffer,
+		 const char **ret_data, const char **ret_data_end);
+
+/**
+ * Decodes `array' and `schema' from the `data' in Arrow IPC format.
+ * Returns 0 on success, -1 on failure (diag is set).
+ */
+int
+arrow_ipc_decode(struct ArrowArray *array, struct ArrowSchema *schema,
+		 const char *data, const char *data_end);
```
Comment (ligurio) [inline src/box/arrow_ipc.h:41]: I would add a fuzzing test for these two functions with roundtrip assertion: `random_buf == encode(decode(random_buf))` because IPROTO is a part of public interface and therefore a part of attack surface.

File: src/box/space.c [hunk 3]
```diff
@@ -1430,6 +1431,21 @@ space_execute_dml(struct space *space, struct txn *txn,
...
+		if (arrow_ipc_decode(&array, &schema, request->arrow,
+				     request->arrow_end) != 0)
+			return -1;
+		assert(array.release != NULL);
+		assert(schema.release != NULL);
+		if (space->vtab->execute_insert_arrow(space, txn, &array,
+						      &schema) != 0)
+			return -1;
+		array.release(&array);
+		schema.release(&schema);
```
Comment (mkostoevr) [inline src/box/space.c:1447]: Should one release the `array` and `schema` prior to the `return -1`?

**Review by mkostoevr — COMMENTED** [review]
LGTM, just a single question.

File: src/box/space.c (hunk 3 above)
Comment (Gumix) [reply src/box/space.c:1447]: Yes, thanks.

File: changelogs/unreleased/gh-10508-iproto_insert_arrow.md (hunk 1 above)
Comment (Gumix) [reply changelogs/unreleased/gh-10508-iproto_insert_arrow.md:4]: Done.

File: src/box/arrow_ipc.h (hunk 2 above)
Comment (Gumix) [reply src/box/arrow_ipc.h:41]: Created issue:
- https://github.com/tarantool/tarantool/issues/10537

File: src/box/space.c (hunk 3 above)
Comment (mkostoevr) [reply src/box/space.c]: Nice approach! I'd do it too verbose.

**Review by mkostoevr — APPROVED** [review]

File: debian/copyright [hunk 4]
```diff
@@ -369,6 +369,14 @@ Files: third_party/arrow/abi.h
 Copyright: 2016-2024 The Apache Software Foundation
 License: Apache-2.0
 
+Files: third_party/nanoarrow/flatcc/*
+Copyright: 2015-2023 Mikkel F. Jørgensen, dvide.com
```
Comment (locker) [inline debian/copyright:373]: We don't use non-ASCII characters in our text files. Please replace `ø` with `o`.

File: third_party/README [hunk 5]
```diff
@@ -66,3 +66,18 @@ How to update arrow/abi.h
...
+
+2. Clone the repository into a temporary location:
+git clone https://github.com/apache/arrow-nanoarrow.git
+
+3. Generate the bundled version of nanoarrow library:
+cd arrow-nanoarrow/
+python ./ci/scripts/bundle.py \
+    --symbol-namespace=Tarantool --with-flatcc --with-ipc \
+    --source-output-dir=tarantool_src/third_party/nanoarrow/ \
+    --include-output-dir=tarantool_src/third_party/nanoarrow/
```
Comment (locker) [inline third_party/README:83]: Why not import it as a submodule and run this command from CMake?

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -225,6 +225,9 @@ set (server_sources
      ${PROJECT_SOURCE_DIR}/third_party/lua-yaml/lyaml.cc
      ${PROJECT_SOURCE_DIR}/third_party/lua-cjson/lua_cjson.c
      ${PROJECT_SOURCE_DIR}/third_party/lua-cjson/strbuf.c
+     ${PROJECT_SOURCE_DIR}/third_party/nanoarrow/flatcc.c
+     ${PROJECT_SOURCE_DIR}/third_party/nanoarrow/nanoarrow.c
+     ${PROJECT_SOURCE_DIR}/third_party/nanoarrow/nanoarrow_ipc.c
```
Comment (locker) [inline src/CMakeLists.txt:230]: Can we build it as a separate library, like zstd?

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/arrow_ipc.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+struct arrow_ipc_buffer *
+arrow_ipc_buffer_new(void);
+
+/** Destroys and deallocates an Arrow IPC buffer. */
+void
+arrow_ipc_buffer_delete(struct arrow_ipc_buffer *ipc_buffer);
+
+/**
+ * Encodes `array' and `schema' into Arrow IPC format. The memory is allocated
+ * by the `ipc_buffer', it must be initialized by `arrow_ipc_buffer_new()'.
```
Comment (locker) [inline src/box/arrow_ipc.h:26]: I don't understand the purpose of this buffer in the API. Maybe, better make it static?

File: src/box/arrow_ipc.c [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+arrow_ipc_buffer_delete(struct arrow_ipc_buffer *ipc_buffer)
+{
+	struct ArrowBuffer *buffer = &ipc_buffer->base;
+	ArrowBufferReset(buffer);
+	TRASH(buffer);
+	free(buffer);
+}
+
+int
+arrow_ipc_encode(struct ArrowArray *array, struct ArrowSchema *schema,
```
Comment (locker) [inline src/box/arrow_ipc.c:37]: I don't see that this function is used anywhere or tested in CE. We shouldn't add dead code.

File: src/box/arrow_ipc.c [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+		diag_set(ClientError, ER_ARROW_IPC_ENCODE,
+			 tt_sprintf("ArrowIpcWriterWriteSchema: %s",
+				    error.message));
+		goto error2;
+	}
+
+	rc = ArrowIpcWriterWriteArrayView(&writer, &array_view, &error);
+	if (rc != NANOARROW_OK) {
+		diag_set(ClientError, ER_ARROW_IPC_ENCODE,
+			 tt_sprintf("ArrowIpcWriterWriteArrayView: %s",
```
Comment (locker) [inline src/box/arrow_ipc.c:99]: IMO better add two error payload field - method and error.

File: src/box/iproto.cc [hunk 10]
```diff
@@ -3664,6 +3665,8 @@ iproto_thread_init_routes(struct iproto_thread *iproto_thread)
 	iproto_thread->dml_route[12] = NULL;
 	/* IPROTO_PREPARE */
 	iproto_thread->dml_route[13] = iproto_thread->sql_route;
+	/* IPROTO_INSERT_ARROW */
+	iproto_thread->dml_route[17] = iproto_thread->process1_route;
```
Comment (locker) [inline src/box/iproto.cc:3669]: Using numbers here looks really ugly. Let's replace them with `IPROTO_XXX` constants while we are at it. Okay to do in a separate commit.

File: src/box/iproto_constants.c [hunk 11]
```diff
@@ -49,6 +49,7 @@ const uint64_t iproto_body_key_map[IPROTO_TYPE_STAT_MAX] = {
 	0,                                                     /* BEGIN */
 	0,                                                     /* COMMIT */
 	0,                                                     /* ROLLBACK */
+	bit(SPACE_ID) | bit(ARROW),			/* INSERT_ARROW */
```
Comment (locker) [inline src/box/iproto_constants.c:52]: Let's use spaces to match other comments.

**Review by locker — CHANGES_REQUESTED** [review]

File: third_party/README (hunk 5 above)
Comment (Gumix) [reply third_party/README]: Switched to submodule with CMake build.

File: debian/copyright (hunk 4 above)
Comment (Gumix) [reply debian/copyright]: Removed this line, because now nanoarrow is imported as a submodule.

File: src/CMakeLists.txt (hunk 6 above)
Comment (Gumix) [reply src/CMakeLists.txt]: Done.

File: changelogs/unreleased/gh-10508-iproto_insert_arrow.md [hunk 12]
```diff
@@ -0,0 +1,4 @@
+## feature/core
+
+* Introduced the new `IPROTO_INSERT_ARROW` request for batch data insertions
+  in the Apache Arrow format (gh-10508).
```
[the remark this answers is not in the document]
Comment (Gumix) [reply changelogs/unreleased/gh-10508-iproto_insert_arrow.md]: > If only memcs supports this new command, is there any point in mentioning in CE release notes?

Removed gh-10508-iproto_insert_arrow.md.

> I think we should implement a generic implementation that would insert an Arrow batch in memtx or vinyl.

I don't have time for it now, maybe later.

File: src/box/arrow_ipc.h (hunk 7 above)
Comment (Gumix) [reply src/box/arrow_ipc.h]: Removed external buffer, changed to:

```c
        /*
         * TODO: It is possible to avoid extra `memcpy()' by switching
         * `ArrowBuffer' to `region_realloc()'.
         */
        char *data = xregion_alloc(region, buffer.size_bytes);
        memcpy(data, buffer.data, buffer.size_bytes);
```

File: src/box/arrow_ipc.c (hunk 8 above)
Comment (Gumix) [reply src/box/arrow_ipc.c:14]: Now it is used by `box_insert_arrow()`, which is not well tested in CE though…

File: src/box/arrow_ipc.c (hunk 9 above)
Comment (Gumix) [reply src/box/arrow_ipc.c]: Changed to:
```c
_(ER_ARROW_IPC_ENCODE, 289, "Failed to encode Arrow IPC data", "method", STRING, "details", STRING) \
_(ER_ARROW_IPC_DECODE, 290, "Failed to decode Arrow IPC data", "method", STRING, "details", STRING) \
```

File: src/box/iproto.cc (hunk 10 above)
Comment (Gumix) [reply src/box/iproto.cc]: Done.

File: src/box/space.c [hunk 13]
```diff
@@ -1344,13 +1344,34 @@ after_old_tuple_lookup:;
...
+		if (space->vtab->execute_insert_arrow(
+				space, txn, array, schema) != 0)
+			return -1;
+		struct region *txn_region = tx_region_acquire(txn);
+		int rc = arrow_ipc_encode(
+				array, schema, txn_region, &request->arrow_ipc,
+				&request->arrow_ipc_end);
+		tx_region_release(txn, TX_ALLOC_SYSTEM);
+		return rc;
+	}
```
Comment (locker) [inline src/box/space.c:1372]: Can we always decode `request->arrow_ipc` and set `arrow_schema` and `arrow_array` in `xrow_decode_dml`?

Can we set `request->arrow_ipc` by encoding `arrow_array` and `arrow_schema` in `txn_add_redo`?

File: src/box/box.cc [hunk 14]
```diff
@@ -4011,6 +4011,19 @@ box_upsert(uint32_t space_id, uint32_t index_id, const char *tuple,
 	return box_process1(&request, result);
 }
 
+API_EXPORT int
+box_insert_arrow(uint32_t space_id, struct ArrowArray *array,
```
Comment (locker) [inline src/box/box.cc:4015]: Please squash the last two patches. Keeping these changes apart make it difficult to review. The new request doesn't make much senses without the new function IMO.

File: src/box/space.c (hunk 13 above)
Comment (Gumix) [reply src/box/space.c]: > Can we always decode request->arrow_ipc and set arrow_schema and arrow_array in xrow_decode_dml?

It's hard to release `array` and `schema` if they are allocated from `xrow_decode_dml()`.

> Can we set request->arrow_ipc by encoding arrow_array and arrow_schema in txn_add_redo?

Done.

File: src/box/box.cc (hunk 14 above)
Comment (Gumix) [reply src/box/box.cc:4015]: Done.

File: src/box/iproto_constants.c [hunk 15]
```diff
@@ -32,23 +32,24 @@
...
+	0,                                                    /* EVAL */
+	bit(SPACE_ID) | bit(OPS) | bit(TUPLE),                /* UPSERT */
+	0,                                                    /* CALL */
+	0,                                                    /* EXECUTE */
+	0,                                                    /* NOP */
+	0,                                                    /* PREPARE */
+	0,                                                    /* BEGIN */
+	0,                                                    /* COMMIT */
+	0,                                                    /* ROLLBACK */
+	bit(SPACE_ID) | bit(ARROW),                           /* INSERT_ARROW */
```
Comment (locker) [inline src/box/iproto_constants.c:52]: Nit: I don't see any point in reformatting the whole array.

**Review by locker — APPROVED** [review]

File: src/box/iproto_constants.c (hunk 15 above)
Comment (Gumix) [reply src/box/iproto_constants.c:52]: It looks ugly when INSERT_ARROW is shifted 1 space to the left to fit 80 characters limit.

**Comment by Gumix** [comment]
Bumped nanoarrow to bring compilation fix.

**Comment by Gumix** [comment]
Rewrote `cmake/BuildNanoarrow.cmake` from `add_subdirectory()` to `ExternalProject_Add()`, because in case of `add_subdirectory()` all build static libs and headers are considered as public, that cause errors during RPM packaging:
```
error: Installed (but unpackaged) file(s) found:
   /usr/include/nanoarrow/common/inline_array.h
   ...
```
